samples/test/clean.js

```diff
@@ -0,0 +1,249 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * This module contains utility functions for removing unneeded, stale, or
+ * orphaned resources from test project.
+ *
+ * Removes:
+ * - Datasets
+ * - Training pipelines
+ * - Models
+ * - Endpoints
+ * - Batch prediction jobs
+ */
+const MAXIMUM_AGE = 3600000 * 24 * 2; // 2 days in milliseconds
+const MAXIMUM_NUMBER_OF_DELETIONS = 10; // Avoid hitting LRO quota
+const TEMP_RESOURCE_PREFIX = 'temp';
+const LOCATION = 'us-central1';
+
+// All AI Platform resources need to specify a hostname.
+const clientOptions = {
+  apiEndpoint: 'us-central1-aiplatform.googleapis.com',
+};
+
+/**
+ * Determines whether a resource should be deleted based upon its
+ * age and name.
+ * @param {string} displayName the display name of the resource
+ * @param {Timestamp} createTime when the resource is created
+ * @returns {bool}
+ */
+function checkDeletionStatus(displayName, createTime) {
+  const NOW = new Date();
+  // Check whether this resources is a temporary resource
+  if (displayName.indexOf(TEMP_RESOURCE_PREFIX) === -1) {
+    return false;
+  }
+
+  // Check how old the resource is
+  const ageOfResource = new Date(createTime.seconds * 1000);
+  if (NOW - ageOfResource < MAXIMUM_AGE) {
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Removes all temporary datasets older than the maximum age.
+ * @param {string} projectId the project to remove datasets from
+ * @returns {Promise}
+ */
+async function cleanDatasets(projectId) {
+  const {DatasetServiceClient} = require('@google-cloud/aiplatform');
+  const datasetServiceClient = new DatasetServiceClient(clientOptions);
+
+  const [datasets] = await datasetServiceClient.listDatasets({
+    parent: `projects/${projectId}/locations/${LOCATION}`,
+  });
+
+  const datasetDeletionOperations = [];
+  for (const dataset of datasets) {
+    const {displayName, createTime, name} = dataset;
+
+    if (checkDeletionStatus(displayName, createTime)) {
+      const deletionOp = datasetServiceClient.deleteDataset({
+        name,
+      });
+      datasetDeletionOperations.push(deletionOp);
+    }
+
+    if (datasetDeletionOperations.length > MAXIMUM_NUMBER_OF_DELETIONS) {
```

Comment:
The reason you're bumping into an LRO issue, I expect, is that `Promise.all` will start all the work in parallel, leading to resource contention.

Instead you could just do:

```js
for (const dataset of datasets) {
  await datasetDeletionOperations(dataset)
}
```

And I think wouldn't need to cap the amount of cleanup you do.

Reply:
Done.